### ISSUE TITLE
Screen coordinates are incorrect when running in Zoomed mode

### ISSUE THREAD

**Issue by jdrm** [calabash/calabash-ios#998]
I am running a couple of calabash-ios tests for my app on a iPhone 6s Plus.
The screen dimensions I get from the console:

    irb(main):006:0> screen_dimensions
    {
        :sample => 1.103448275862069,
        :height => 2001,
         :width => 1125,
         :scale => 3
    }

I am trying to touch a button in a TOU screen:

[![enter image description here][1]][1]

  [1]: http://i.stack.imgur.com/eKuJAl.png

If I query for the Accept button what I get is:

    irb(main):007:0> query("button")
    [
        [0] {
                              "id" => nil,
                     "description" => "<_UIToolbarNavigationButton: 0x12de3e230; frame = (0 6.66667; 54 30); opaque = NO; layer = <CALayer: 0x12de04990>>",
                           "label" => "Accept",
                           "frame" => {
                     "y" => 6.67,
                     "x" => 0,
                 "width" => 54,
                "height" => 30
            },
            "accessibilityElement" => true,
                           "value" => nil,
                           "alpha" => 1,
                         "enabled" => true,
                         "visible" => 1,
                        "selected" => false,
                           "class" => "_UIToolbarNavigationButton",
                            "rect" => {
                       "y" => 629.67,
                "center_x" => 359.72,
                "center_y" => 711.36,
                       "x" => 299,
                   "width" => 54,
                  "height" => 30
            }
        }
    ]

Problem is that if I try to touch the button what I get this error:

    irb(main):008:0> touch("button")
    RuntimeError: uia action failed because: Input: uia.tapOffset('{:x 359.72, :y 711.36}')
    . Error: VerboseError: Unable to tap-offset: {:y 711.36, :x 359.72}  [native code]

It tries to touch the (center_y, center_x) but those coordinates are apparently out of the bounds of the screen.

It works just fine if I instead try to touch passing some valid screen coordinates:

    irb(main):009:0> touch(nil, :offset => {:x => 300, :y => 630})
    true

Apparently the coordinates calculated are not correct. If 2001 is the max height and the scale is 3 then the calculated y_center coordinate 711.36 is apparently out of bounds.

I already tried running on simulators (including iPhone 6s Plus simulator) and it works just fine. I also tried running on a physical iPhone 5c and it also works fine. I was unable to find documentation indicating that I should pass some additional parameter when running on a physical iPhone 6s Plus for the coordinates to be calculated correctly.

My env info:

    Jimmys-MacBook-Pro:/ jdrm$  xcode-select --print-path
    /Applications/Xcode.app/Contents/Developer

    Jimmys-MacBook-Pro:/ jdrm$ xcodebuild -version
    Xcode 7.2
    Build version 7C68

    Jimmys-MacBook-Pro:/ jdrm$ calabash-ios version
    0.18.0

    Jimmys-MacBook-Pro:/ jdrm$ curl http://172.22.21.23:37265/version
    {"device_family":"iPhone","outcome":"SUCCESS","server_port":37265,"simulator_device":"","simulator":"","app_name":"","app_version":"2.3.1.b1","screen_dimensions":{"sample":1.103448275862069,"height":2001,"width":1125,"scale":3},"git":{"revision":"97a2e3e","remote_origin":"git@github.com:calabash\/calabash-ios-server.git","branch":"master"},"device_name":"Jimmy's iPhone","4inch":false,"app_id":"XXX","form_factor":"iphone 6+","system":"iPhone8,2","version":"0.18.0","iOS_version":"9.2.1","short_version_string":"2.3.1.b1","ios_version":"9.2.1","iphone_app_emulated_on_ipad":false,"model_identifier":"iPhone8,2","app_base_sdk":"iphoneos9.2"}

**Comment by TeresaP** [comment]
That sounds like the problem we're having on our iPad Pro

**Comment by jmoody** [comment]
@jdrm Please please please do not cross post.  I see you have the exact issue in Stack Overflow.  We monitor all the public channels.  Cross posting doubles the work for us.  We would appreciate it if you closed the Stack Overflow question.

>  I was unable to find documentation indicating that I should pass some additional parameter when running on a physical iPhone 6s Plus for the coordinates to be calculated correctly.

No.  There is no additional parameter required.

The problem here is that the app is being displayed in Scaled Mode either because:

1. Zoom mode is on
2. The app is not optimized for the larger screens
  * Does not have @3x assets
  * Does not have the correct launch images

In either case Calabash should be able to deduce the correct coordinates.

Can you share this app with us?  All of our testing apps optimized for larger screens.

Is this the only button that cannot be touched?  Is this view being presented modally?

**Comment by jdrm** [comment]
Sorry for the cross-posting. I just deleted the question in Stack Overflow.
My device was indeed with Zoom mode on. I just tried disabling it and my test were able to run without problems. This is not the only button that cannot be touched this happens for all the screen of my app. I also just reproduced the problem using this sample app:
https://github.com/calabash/ios-smoke-test-app

    irb(main):007:0> query("UITabBarButton marked:'Date Picker'")
    [
        [0] {
                              "id" => nil,
                     "description" => "<UITabBarButton: 0x146f82c30; frame = (302 1; 71 48); opaque = NO; layer = <CALayer: 0x146d538c0>>",
                           "label" => "Date Picker",
                           "frame" => {
                     "y" => 1,
                     "x" => 302,
                 "width" => 71,
                "height" => 48
            },
            "accessibilityElement" => true,
                           "value" => nil,
                           "alpha" => 1,
                         "enabled" => true,
                         "visible" => 1,
                        "selected" => false,
                           "class" => "UITabBarButton",
                            "rect" => {
                       "y" => 619,
                "center_x" => 372.41,
                "center_y" => 709.52,
                       "x" => 302,
                   "width" => 71,
                  "height" => 48
            }
        }
    ]


When trying to touch the date picker:

    irb(main):009:0> touch("UITabBarButton marked:'Date Picker'")
    RuntimeError: uia action failed because: Input: uia.tapOffset('{:x 372.41, :y 709.52}')
    . Error: VerboseError: Unable to tap-offset: {:y 709.52, :x 372.41}  [native code]

**Comment by jmoody** [comment]
> touch("UITabBarButton marked:'Date Picker'")

Is this in Zoomed mode?

Did you call `start_test_server_in_background` before calling `touch`?

This app is tested nightly in CI against simulators and devices.

**Comment by jdrm** [comment]
Yes this is in zoomed mode.
Yes I called start_test_server_in_background before calling touch. I pasted the result of querying for that same element.

I was also able to successfully touch the mentioned element by running:
    
    irb(main):013:0> touch(nil, :offset => {:x => 320, :y => 630})
    true

So the calabash-ios is working but apparently failing to calculate the correct offset

**Comment by jmoody** [comment]
I have marked this as a bug.  Until this is fixed, the workaround is to:

1. Don't run in Zoomed mode
2. Update your application to be optimized for the larger screen sizes
  * provide @3x assets
  * provide the necessary launch images

**Comment by jdrm** [comment]
Just to understand this for workaround #2 you mention I should assume that the ios-smoke-test-app is not optimized for the larger screen sizes?

**Comment by jmoody** [comment]
CalSmokeApp _is_ optimized for larger screen sizes.

**Comment by jmoody** [comment]
Will have to wait for 0.19.0 - the 0.18.2 server has been released.

The workaround is to:

1. Optimize your app for the 6 and 6 plus form-factors (as mentioned above)
2. Test with your application in Standard (not Zoomed) mode

**Comment by jmoody** [comment]
I have fixed this in the develop branch of the server.  The change will be available in 0.19.0.

The only caveat is this:

```
This PR fixes the touch coordinates for all these cases except one: Non-optimized apps in Zoomed display mode on iPhone 6 Plus physical devices.

There are two workarounds:

Update your app to be optimized for the iPhone 6* screen sizes
Test in Standard display mode
```
